open_spiel/python/algorithms/deep_cfr_tf2.py

```diff
@@ -0,0 +1,614 @@
+      self._add_to_strategy_memory(state.information_state_tensor(other_player), self._iteration,
+              strategy, state.legal_actions_mask(other_player))
+      return self._traverse_game_tree(state.child(sampled_action), player)
+
+  def _sample_action_from_advantage(self, state, player):
+    """Returns an info state policy by applying regret-matching.
+
+    Args:
+      state: Current OpenSpiel game state.
+      player: (int) Player index over which to compute regrets.
+    Returns:
+      1. (np-array) Advantage values for info state actions indexed by action.
+      2. (np-array) Matched regrets, prob for actions indexed by action.
+    """
+    info_state = tf.constant(state.information_state_tensor(player), dtype=tf.float32)
+    legal_actions_mask = tf.constant(state.legal_actions_mask(player), dtype=tf.float32)
+    advs = self._adv_networks[player]((tf.expand_dims(info_state, axis=0), legal_actions_mask), training=False)[0]
+    advs = advs * legal_actions_mask
+    advantages = tf.maximum(advs, 0)
+    cumulative_regret = tf.reduce_sum(advantages)
+    if cumulative_regret > 0:
+      matched_regrets = advantages / cumulative_regret
+    else:
+      matched_regrets = tf.one_hot(tf.argmax( tf.where(legal_actions_mask==1, advs, -10e20)), self._num_actions)
+    return advantages.numpy(), matched_regrets.numpy()
+
+  def action_probabilities(self, state):
+    """Returns action probabilities dict for a single batch."""
+    cur_player = state.current_player()
+    legal_actions = state.legal_actions(cur_player)
+    legal_actions_mask = tf.constant(state.legal_actions_mask(cur_player), dtype=tf.float32)
+    info_state_vector = tf.constant(state.information_state_tensor(), dtype=tf.float32)
+    if len(info_state_vector.shape) == 1:
+      info_state_vector = tf.expand_dims(info_state_vector, axis=0)
+    probs = self._policy_network((info_state_vector, legal_actions_mask), training=False)
+    probs = probs.numpy()
+    return {action: probs[0][action] for action in legal_actions}
+
+  def _get_advantage_dataset(self, player):
+    """returns the collected regrets for the given player as a dataset"""
+    random.shuffle(self._advantage_memories[player]._data)
+    data = tf.data.Dataset.from_tensor_slices(self._advantage_memories[player]._data)
+    data = data.shuffle(100000)
+    data = data.batch(self._batch_size_advantage, drop_remainder=True)
+    data = data.map(self._deserialize_advantage_memory)
+    data = data.repeat()
+    data = data.prefetch(10)
+    return data
+
+  def _learn_advantage_network(self, player):
+    """Compute the loss on sampled transitions and perform a Q-network update.
+
+    If there are not enough elements in the buffer, no loss is computed and
+    `None` is returned instead.
+
+    Args:
+      player: (int) player index.
+    Returns:
+      The average loss over the advantage network of the last batch.
+    """
+    
```

Comment:
I needed to put the following here, or I got a Value error crash on line 575 below (taken from deep_cfr.py):

```
    if self._batch_size_advantage:
      if self._batch_size_advantage > len(self._advantage_memories[player]):
        ## Skip if there aren't enough samples
        return None
```

Reply:
I've instead changed the _get_advantage_dataset(...) method to work if len(self._advantage_memories[player]) < self._batch_size_advantage